File: bindings/python/sop/browser.py
import os
import sys
import platform
import urllib.request
import subprocess
import stat
from pathlib import Path

# Try to import version from the package, fallback to latest if running standalone
try:
    from . import __version__
except ImportError:
    __version__ = "latest"

GITHUB_REPO = "sharedcode/sop"
BINARY_NAME = "sop-httpserver"

def get_platform_info():
    system = platform.system().lower()
    machine = platform.machine().lower()
    
    if system == "darwin":
        os_name = "darwin"
    elif system == "linux":
        os_name = "linux"
    elif system == "windows":
        os_name = "windows"
    else:
        raise RuntimeError(f"Unsupported operating system: {system}")

    if machine in ["x86_64", "amd64"]:
        arch = "amd64"
    elif machine in ["arm64", "aarch64"]:
        arch = "arm64"
    else:
        raise RuntimeError(f"Unsupported architecture: {machine}")
        
    return os_name, arch

def get_binary_path():
    home = Path.home()
    sop_dir = home / ".sop" / "bin"
    sop_dir.mkdir(parents=True, exist_ok=True)
    
    ext = ".exe" if platform.system() == "Windows" else ""
    return sop_dir / (BINARY_NAME + ext)

def download_binary(os_name, arch, target_path):
    # Use the package version to find the matching release
    # If version is 'latest', use the latest endpoint
    if __version__ == "latest":
        base_url = f"https://github.com/{GITHUB_REPO}/releases/latest/download"
    else:
        # Use a prefixed tag to distinguish from the main Go project releases (e.g. v5.2)
        tag = f"sop4py-v{__version__}"
        base_url = f"https://github.com/{GITHUB_REPO}/releases/download/{tag}"

    filename = f"{BINARY_NAME}-{os_name}-{arch}"
    if os_name == "windows":
        filename += ".exe"
        
    url = f"{base_url}/{filename}"
    
    print(f"Downloading SOP Data Browser ({__version__}) from {url}...")
    try:
        urllib.request.urlretrieve(url, target_path)
        print("Download complete.")
        
        # Make executable
        st = os.stat(target_path)
        os.chmod(target_path, st.st_mode | stat.S_IEXEC)
        
    except Exception as e:
        print(f"Error downloading binary: {e}")
        if target_path.exists():
            target_path.unlink()
        sys.exit(1)

def check_version(binary_path):
    if __version__ == "latest":
        return True
        
    try:
        # Run binary with --version
        result = subprocess.run([str(binary_path), "--version"], capture_output=True, text=True)
        if result.returncode != 0:
            return False
            
        # Output format: "SOP Data Browser v2.0.32"
        output = result.stdout.strip()
        if f"v{__version__}" in output:
            return True
            
        print(f"Version mismatch: Found {output}, expected v{__version__}")
        return False
    except Exception:
        return False

def main():
    try:
        os_name, arch = get_platform_info()
        binary_path = get_binary_path()
        
        if not binary_path.exists() or not check_version(binary_path):
            if binary_path.exists():
                print("Updating SOP Data Browser...")
                try:
                    binary_path.unlink()
                except Exception:
                    pass # Ignore if we can't delete, download might overwrite or fail
            else:
                print(f"SOP Data Browser not found at {binary_path}")
            
            download_binary(os_name, arch, binary_path)
            
        print(f"Starting SOP Data Browser...")
        
        # Pass through any arguments
        args = [str(binary_path)] + sys.argv[1:]
        
        # Replace current process with the binary
        if platform.system() == "Windows":
            subprocess.run(args)
        else:
            os.execv(str(binary_path), args)
            
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: bindings/csharp/Sop.Examples/Browser.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Net.Http;
using System.Runtime.InteropServices;
using System.Threading.Tasks;

namespace Sop.Examples;

public static class Browser
{
    private const string GithubRepo = "sharedcode/sop";
    private const string BinaryName = "sop-browser";
    // TODO: Sync this version with the build system
    private const string Version = "2.0.40"; 

    public static async Task Run()
    {
        try
        {
            var (osName, arch) = GetPlatformInfo();
            var binaryPath = GetBinaryPath();

            if (!File.Exists(binaryPath))
            {
                Console.WriteLine($"SOP Data Browser not found at {binaryPath}");
                await DownloadBinary(osName, arch, binaryPath);
            }

            Console.WriteLine("Starting SOP Data Browser...");
            
            var psi = new ProcessStartInfo
            {
                FileName = binaryPath,
                UseShellExecute = false
            };
            
            using var process = Process.Start(psi);
            process?.WaitForExit();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error: {ex.Message}");
        }
    }

    private static (string os, string arch) GetPlatformInfo()
    {
        string os = "";
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) os = "windows";
        else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) os = "linux";
        else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) os = "darwin";
        else throw new PlatformNotSupportedException("Unsupported OS");

        string arch = RuntimeInformation.ProcessArchitecture switch
        {
            Architecture.X64 => "amd64",
            Architecture.Arm64 => "arm64",
            _ => throw new PlatformNotSupportedException($"Unsupported Architecture: {RuntimeInformation.ProcessArchitecture}")
        };

        return (os, arch);
    }

    private static string GetBinaryPath()
    {
        var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        var sopDir = Path.Combine(home, ".sop", "bin");
        Directory.CreateDirectory(sopDir);

        var ext = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? ".exe" : "";
        return Path.Combine(sopDir, BinaryName + ext);
    }

    private static async Task DownloadBinary(string os, string arch, string targetPath)
    {
        // Use the Sop4CS release tag
        string tag = $"Sop4CS-v{Version}"; 
        
        string baseUrl = $"https://github.com/{GithubRepo}/releases/download/{tag}";
        string filename = $"{BinaryName}-{os}-{arch}";
        if (os == "windows") filename += ".exe";

        string url = $"{baseUrl}/{filename}";

        Console.WriteLine($"Downloading SOP Data Browser ({Version}) from {url}...");

        using var client = new HttpClient();
        var response = await client.GetAsync(url);
        response.EnsureSuccessStatusCode();

        using var fs = new FileStream(targetPath, FileMode.Create);
        await response.Content.CopyToAsync(fs);
        
        if (os != "windows")
        {
            try
            {
                Process.Start("chmod", $"+x {targetPath}")?.WaitForExit();
            }
            catch { /* ignore */ }
        }
        
        Console.WriteLine("Download complete.");
    }
}
